pkg/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,370 @@
+	// may become unreferenced and need to be deleted; this will be handled async
+	// in the manageCache loop.
+
+	// make changes to the watches
+	if err := c.replaceWatchSet(ctx); err != nil {
+		return fmt.Errorf("error watching new gvks: %w", err)
+	}
+
+	return nil
+}
+
+// replaceWatchSet looks at the specifiedGVKs and makes changes to the registrar's watch set.
+// assumes caller has lock.
+// replaceWatchSet may error out and that error is retryable.
+func (c *CacheManager) replaceWatchSet(ctx context.Context) error {
+	newWatchSet := watch.NewSet()
+	newWatchSet.Add(c.gvksToSync.GVKs()...)
+
+	// record any gvks that need to be deleted from the opa cache.
+	c.gvksToDeleteFromCache.AddSet(c.watchedSet.Difference(newWatchSet))
+
+	var innerError error
+	c.watchedSet.Replace(newWatchSet, func() {
+		// *Note the following steps are not transactional with respect to admission control
+
+		// Important: dynamic watches update must happen *after* updating our watchSet.
+		// Otherwise, the sync controller will drop events for the newly watched kinds.
+		innerError = c.registrar.ReplaceWatch(ctx, newWatchSet.Items())
+	})
+
+	return innerError
+}
+
+// RemoveSource removes the watches of the GVKs for a given aggregator.Key.
+// It errors out if there is an issue removing the Key internally or replacing the watches.
+// Consumers are encouraged to retry on error.
+func (c *CacheManager) RemoveSource(ctx context.Context, sourceKey aggregator.Key) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if err := c.gvksToSync.Remove(sourceKey); err != nil {
+		return fmt.Errorf("internal error removing source: %w", err)
+	}
+
+	// make changes to the watches
+	if err := c.replaceWatchSet(ctx); err != nil {
+		return fmt.Errorf("error removing watches for source %v: %w", sourceKey, err)
+	}
+
+	return nil
+}
+
+// ExcludeProcesses swaps the current process excluder with the new *process.Excluder.
+// It's a no-op if the two excluders are equal.
+func (c *CacheManager) ExcludeProcesses(newExcluder *process.Excluder) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.processExcluder.Equals(newExcluder) {
+		return
+	}
+
+	c.processExcluder.Replace(newExcluder)
+	// there is a new excluder which means we need to schedule a wipe for any
+	// previously watched GVKs to be re-added to get a chance to be evaluated
+	// for this new process excluder.
+	c.excluderChanged = true
+}
+
+func (c *CacheManager) AddObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	gvk := instance.GroupVersionKind()
+
+	isNamespaceExcluded, err := c.processExcluder.IsNamespaceExcluded(process.Sync, instance)
+	if err != nil {
+		return fmt.Errorf("error while excluding namespaces for gvk: %+v: %w", gvk, err)
+	}
+
+	// bail because it means we should not be
+	// syncing this gvk's objects as it is namespace excluded.
+	if isNamespaceExcluded {
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
```

Comment:
Should we also cancel the expectation if the watch set does not contain this GVK?

Reply:
That would lead to over-cancelling of expectations, since expectations are gathered by the readiness tracker (which lists them separately from the config/SyncSet controllers). In an extreme case, this could bypass expectations altogether.

Are we canceling expectations in the config controller still?

https://github.com/open-policy-agent/gatekeeper/blob/ea842cd06fbd85049032947fe81952ece7167a2d/pkg/controller/config/config_controller.go#L388-L393

We'll need to have some fancier logic in the readiness tracker once expectations can come from two bodies, We'll probably need to do reference counting.

Come to think of it. There is a slight race condition in the readiness tracker, where if a resource is deleted before the informer has synced, then it's possible for us to miss delete events, causing g8r to hang. We could wait for informers to sync, but that wouldn't help with dynamic watches (syncs, constraints, etc.). Might need a way to prune stale expectations (i.e. underlying object has been deleted), if readiness hasn't been achieved by a certain time (we'll need to be careful there, since we're hitting the API server directly).

Reply:
WRT over-canceling, we'd probably not see it in practice, since we should only see an object if it's already being watched.

However, note that the above code cancels expectations for an entire GVK, not just a single object.

Reply:
> Are we canceling expectations in the config controller still?

no we are not canceling expectations in the config controller anymore.

> We'll need to have some fancier logic in the readiness tracker once expectations can come from two bodies, We'll probably need to do reference counting.

ack, I will keep that in mind when making the tracker changes.

It sounds like there isn't anything left to do for this comment so I am going to resolve it. Y'all please let me know otherwise.